File: tl1.py
import sys
import os

varDic = {}
    
#main文
def main():
    loadText()
    args = sys.argv
    openFile = args[1]
    blankWord = ['\n', ' ', '\t', ';']

    with open(openFile) as f:
        for line in f:
            line = line.strip('\n')
            for word in blankWord:
                if line.find(word):
                    continue
            if line[1] == '=':
                if line[3] == ';':
                    varDic[line[0]] = line[2]
                elif line[3] == '+' and line[5] == ';':
                    varDic[line[0]] = getNumber(line[2]) + getNumber(line[4])
                elif line[3] == '-' and line[5] == ';':
                    varDic[line[0]] = getNumber(line[2]) + getNumber(line[4])
                else:
                    err(line)
            elif line[0] == 'p' and line[1] == 'r' and\
                    line[5] == ' ' and line[7] == ';':
                print("{}".format(varDic[line[6]]))
            else:
                err(line)


#プログラムファイルを読み込む
def loadText():
    args = sys.argv
    if len(args) < 2:
        print("useage {} program-file".format(args[0]))
        sys.exit()
    if not os.path.exists(args[1]):
        noFileErr(args[1])
    with open(args[1]) as f:
        if f == 0:
            print("fopen error : {}".format(args[1]))
            sys.exit()

#数字ならばint型で、変数ならば変数の値を返す
def getNumber(txt):
    if txt.isnumeric():
        return int(txt)
    return int(varDic[txt])


#プログラムの構文エラー
def err(strLine):
    print("syntax error : {}".format(strLine))
    sys.exit()


#ファイルが存在しないエラー
def noFileErr(fileStr):
    print("no file error : {}".format(fileStr))
    sys.exit()

if __name__ == "__main__":
    main()

File: tl1.c
#include<stdio.h>
#include<stdlib.h>

void loadText(int argc, const char **argv, unsigned char *t, int siz) {
    FILE *fp;
    int i;
    if(argc < 2) { //引数が少ないのでエラー表示して終了
        printf("usage>%s program-file\n", argv[0]);
        exit(1);
    }
    fp = fopen(argv[1], "rt"); //テキストモードでファイルを開く
    if(fp == 0) { //ファイルがひらけなかった
        printf("fopen error : %s\n", argv[1]);
        exit(1);
    }
    i = fread(t, 1, siz - 1, fp);
    fclose(fp);
    t[i] = 0; //終端マークを書いておく
}

int var[256]; //変数
unsigned char txt[10000]; //ソースコード
int pc = 0; //プログラムカウンタ

int getNumber(unsigned char c) {
    if('0' <= c && c <= '9')
        return c - '0'; //1桁の定数
    return var[c]; //1文字の変数名
}

int main(int argc, const char **argv) {
    int i, pc0;
    loadText(argc, argv, txt, 10000);
    for(;;) {
        pc0 = pc;
        if(txt[pc] == 0) //ファイル終端
            exit(0);
        if(txt[pc] == '\n' || txt[pc] == ' ' || txt[pc] == '\t' ||
            txt[pc] == ';') {
            pc++;
            continue;        
        }
        if(txt[pc + 1] == '=') { //2文字目が"="
            i = txt[pc];
            if(txt[pc + 3] == ';') { //単純代入
                var[i] = getNumber(txt[pc + 2]);
            } else if(txt[pc + 3] == '+' && txt[pc + 5] == ';') { //加算
                var[i] = getNumber(txt[pc + 2]) + getNumber(txt[pc + 4]);
            } else if(txt[pc + 3] == '-' && txt[pc + 5] == ';') { //減纂
                var[i] = getNumber(txt[pc + 2]) - getNumber(txt[pc + 4]);
            } else
                goto err;
        } else if(txt[pc] == 'p' && txt[pc + 1] == 'r' &&
                txt[pc + 5] == ' ' && txt[pc + 7] == ';') { //print
            printf("%d\n", var[txt[pc + 6]]);
        } else
            goto err;
        while(txt[pc] != ';')
            pc++;
    }
err:
    printf("syntax error : %.10s\n", &txt[pc0]);
    exit(1);
}